### PR TITLE
Review of HOPE-3

### PR DESCRIPTION
**Pull request by jay-tyler**
Some minor API refactors to consider.

Mostly focused on code blocks, but managed to update some of the surrounding text.

### CODE REVIEW

File: all/HOPE-3--Dependency-Injection.md [hunk 1]
```diff
@@ -78,45 +78,52 @@ It should also be possible to specify different dependency providers for differe
...
+    @hug.cli.provide
     def mysql():
         ...
 ```
 
 If you only want to enable provider usage locally, and don't want to allow it to be overridden, you can explicitly set it to stay unnamed:
 
 ```python
-    @hug.provides(name=False)
+    @hug.provide(name=False)
```
Comment (jay-tyler) [inline all/HOPE-3--Dependency-Injection.md:108]: Is `name` necessary in addition to the `api` kwarg? You could maybe pass `__name__` to API instead.

Just thinking of opportunities to simplify the interface for testing.

File: all/HOPE-3--Dependency-Injection.md [hunk 2]
```diff
@@ -78,45 +78,52 @@ It should also be possible to specify different dependency providers for differe
...
+via an eponymous argument
 
 ```python
-from hug import requires
 
-@hug.http()
-def hello_world(mysql=requires('mysql')):
+@hug.http
+@hug.require.mysql
+def hello_world(mysql):
```
Comment (Peter200lx) [inline all/HOPE-3--Dependency-Injection.md:122]: Personally I like the `parameter=requires('dependency')` format more than having a new magic parameter from a decorator. I feel like this explicit function call pattern is more immediately understandable as to what is happening.

File: all/HOPE-3--Dependency-Injection.md [hunk 3]
```diff
@@ -23,10 +23,10 @@ This system would be built with the intention of replacing Hug's current `direct
 ## Proposed Dependency Injection System
 
 This HOPE proposes a new dependency injection system with an aim for clarity and reuse.
-The API of the new system would consist of a `provides` decorator for defining available dependencies, and a `requires` function for using them:
+The API of the new system would consist of a `provide` decorator for defining available dependencies, and a `requires` function for using them:
```
Comment (Peter200lx) [inline all/HOPE-3--Dependency-Injection.md:26]: I don't have any strong feelings behind using the plural or singular form here, Out of curiosity could you clarify why you prefer one over the other?

File: all/HOPE-3--Dependency-Injection.md (hunk 2 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:122]: Here are the motivating thoughts I had in refactoring this:
- get rid of string value connascence
- also, in contrast to string connascence, we could get IDE tab over completion (implementation and IDE dependent)
- would have functional closure to accommodate monkey patching on dynamic changes. The string default form evaluates on import time only
- get rid of a "fake arg". It's a little odd to use a signature to define an arg which is only ever intended to take a default value. Overriding the default would be a violation of our intended contract, but this is at odds with how Python generally models data.
- pytest has high traction and a pretty good "magic" API. I think we can expect users to understand magic that would be similar or the same to pytest. This is close to the pytest fixture magic, although using the decorator makes this pattern more explicit than the pytest fixture pattern

Happy to discuss further!

File: all/HOPE-3--Dependency-Injection.md (hunk 3 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:26]: Require hints towards active voice, whereas requires hints towards passive. e.g. "It requires" vs "I require".

Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:26]: I was talking about the `provide` vs `provides` more than the use (require) in this case but I can see the argument being similar. That being said I think
```python
@hug.provide('identifier')
def function(...
```
is a little bit more open to interpretation to what is doing the action. Is Hug providing something to the decorated function or is the decorated function providing something to the API?
On the other hand
```python
@hug.provides("identifier")
def function(...
```
to  me reads more clearly as the function providing `identifier` to the API.

However as I stated in the first comment I don't have strong feelings on this point, I'm fine with either form.

File: all/HOPE-3--Dependency-Injection.md (hunk 2 above)
Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:122]: I think points 3 and 4 (1 indexed) counteract each-other a bit as providing a parameter that defaults to a dependency is an acceptable behavior for test cases. A test case could either provide the arguments that the dependency requires and leave parameter alone or it could ignore the dependency injection arguments and directly override the parameter with the data the dependency would provide. (This does make me think of a separate point for this HOPE, there should be a warning if a dependency parameter is provided as well as the parameter to which the dependency return is assigned)

For point 1 I can see the value in reducing the amount of "strings as code",  but one way or the other the named `identifier` has to be referenced indirectly through the API. To pull in point 2 I think we could potentially have `requires` be a class such that you could call:
```python
@hug.http()
def hello_world(mysql=requires.mysql()):
```
I actually like this form a bit better than the initial suggestion at line 31 in that it is more clear that `*args, **kwargs` goes directly to the dependency function.

Per point 5 I know pytest uses the magic parameter name, but I've seen it cause confusion for multiple new coders, and personally I think testing code has different behavior standards as by its nature it is performing meta logic. That being said HUG itself has liberally used this format in the past so I don't consider it as unacceptable when designing new HUG features. I think it is reasonable behavior, but that's not the same saying it is the cleanest form for designing new functionality as in this HOPE.

I'm quite open to changing my thoughts here, just want to make sure I'm explaining myself clearly. I definitely value having multiple viewpoints on the HOPE!

File: all/HOPE-3--Dependency-Injection.md (hunk 3 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:26]: The default behavior is to use the wrapped function name, I think overriding default behavior should require a named kwarg, e.g.

```
@hug.provide(alias="heres_a_name")
def function(...)
...

File: all/HOPE-3--Dependency-Injection.md (hunk 2 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:122]: I dunno, it's apples to oranges on the magic end. I don't think "magic" is necessarily a bad thing, but it's probably better to leverage magic via a decorator when it's needed. It's explicit and well understood that a decorator is allowed to change callable flow. Also it's the only kind of magic we see provided in the standard library. Thinking of `functools.total_ordering` and `contextlib.context_manager`. Both very "magic" but in a fairly explicit and understandable way.

Thinking the default arg approach
- Raises questions about control flow (why a default arg? and who is calling hello_world? and where is the contract of hello_world specified fully? )
- Raises questions about what requires.mysql is doing
- Is harder to remember as a pattern

Thinking the decorator approach
- Raises a question about the `hug.provide` decorator and how it works

Not totally understanding your point about points 3 and 4 counteracting each other. Point 3 is the import time issue, and I'm not sure how you'd get around it. It's going to evaluate as an expression at import time, and the only way to get around it is to return a callable as a result of `requires.mysql()`, e.g. `requires.mysql()()` would give the value. But that would be fairly confusing to end users IMO.

I think both approaches raise questions for beginners. Most interested in the approach that helps people pose a question to the right place (in code, ideally), and is easiest to remember as a pattern once used.

Also - what unit test framework is using default arg setting as an injection mechanism? I'd like to take a look.

File: all/HOPE-3--Dependency-Injection.md (hunk 3 above)
Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:26]: Fair enough, I had forgotten that `name=` was an optional parameter and that it would use the wrapped function's name if not provided. I'll need to think more on if I agree with requiring that it is a named kwarg. Regardless in the case where the decorator is setting up a dependency injection named from the function I agree `@hug.provide()` sounds a bit better. This change seems sensible to me.

File: all/HOPE-3--Dependency-Injection.md (hunk 2 above)
Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:122]: As I've thought this through I agree the level of magic is fairly similar in either pattern. I agree that using a decorator to change the call-time behavior of the function is the more recognizable pattern. The decorator way the HUG code doesn't need to reach inside the function using require to associate the parameter with the dependency injection.

I'm curious to get @timothycrosley's thoughts, but my understanding of the `def function(db = requires('mysql'))` is that a callable was stored in the default parameter for `db` and that HUG would look for the callable in the function's parameter defaults when handing off execution to the endpoint. Anyhow I agree that it is not an immediately recognizable behavior pattern.

I'm not aware of any unit test framework using default arguments to parameters as an injection mechanism. I was talking about pytest same as your first comment's 5th bullet point. The part I particularly found confusing in pytest is that it had to introspect the functions you wrote to determine the name of parameters and match fixtures to those names. However when there's a decorator directly attached to the function (and especially if the decorator name matches the parameter name) it reduces the scope of investigation to track down the source of data to the parameter.

tl;dr: I agree with moving to a decorator for `require`

File: all/HOPE-3--Dependency-Injection.md [hunk 4]
```diff
@@ -58,18 +58,18 @@ def my_endpoint(timer: hug.directives.timer):
 
 ### Defining a dependency
 
-Defining a dependency will happen via a new hug `provides` decorator:
+Defining a dependency will happen via a new hug `provide` decorator:
 
 ```python
-    @hug.provides()
+    @hug.provide
```
Comment (Peter200lx) [inline all/HOPE-3--Dependency-Injection.md:64]: Hug's decorators are not currently optionally callable. To support that requires additional changes to HUG's code. For example:
https://chase-seibert.github.io/blog/2013/12/17/python-decorator-optional-parameter.html#optional-parameters
https://realpython.com/primer-on-python-decorators/#both-please-but-never-mind-the-bread
https://github.com/dabeaz/python-cookbook/blob/master/src/9/defining_a_decorator_that_takes_an_optional_argument/example.py

I think it might be neat to explore potentially adding support to HUG, but that should be a discussion for a separate HOPE, not edited in passing in this one.

**Comment by Peter200lx** [comment]
I think some of the discussion from this PR has come from there being different understandings of the intended implementation of the new interfaces discussed in this HOPE. @timothycrosley could you add your thoughts on the high level implementation overview for this HOPE? (Both what you were thinking as regards implementation and also if that has had any alterations from the discussion here)

File: all/HOPE-3--Dependency-Injection.md (hunk 4 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:64]: I'm borrowing the optional callability from the pytest fixture pattern.

I'd consider this an implementation detail, which a HOPE probably shouldn't be concerned with.

**Comment by jay-tyler** [comment]
> I think some of the discussion from this PR has come from there being different understandings of the intended implementation of the new interfaces discussed in this HOPE. @timothycrosley could you add your thoughts on the high level implementation overview for this HOPE? (Both what you were thinking as regards implementation and also if that has had any alterations from the discussion here)

100% agree that more viewpoints would be welcome and helpful. But when you say "implementation" are you thinking of the code required to make these APIs a reality? My understanding is that these are design docs, which probably shouldn't be concerned with implementation details outside of understanding possible execution dependencies. It's totally possible I'm misunderstanding what you mean by "implementation" though.

File: all/HOPE-3--Dependency-Injection.md (hunk 4 above)
Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:64]: You can see how PyTest had to check for the `callable` nature of the first parameter here:
https://github.com/pytest-dev/pytest/blob/master/src/_pytest/fixtures.py#L1036-L1041

Part of the reason why I'm highlighting this here is that you also converted `@hug.http()` below to `@hug.http` which is out of scope for this HOPE I believe. Converting a decorator to be optionally callable adds some constraints on the nature of the parameters passed into the decorator. If those constraints are acceptable I think it would make sense to apply the pattern more broadly throughout HUG. I just want to make sure the tradeoffs are considered before changing the behavior.

**Comment by Peter200lx** [comment]
> But when you say "implementation" are you thinking of the code required to make these APIs a reality?

I don't mean the full in-code implementation. By "high level implementation overview" I mean specifying more details about how HUG would store the `provides` injections into the global or interface local scopes. And more importantly how the `requires('identifier')`  would be activated upon an API call hitting the specific endpoint/function.

> My understanding is that these are design docs, which probably shouldn't be concerned with implementation details outside of understanding possible execution dependencies.

I definitely agree we don't want specifications that are too detailed because the point of a HOPE is to be a design doc. However when there's confusion as to the impact of the changes when two people read the same HOPE and are picturing a different design then that might be a sign that more detail is needed in the design doc.

File: all/HOPE-3--Dependency-Injection.md (hunk 4 above)
Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:64]: > Converting a decorator to be optionally callable adds some constraints on the nature of the parameters passed into the decorator. If those constraints are acceptable I think it would make sense to apply the pattern more broadly throughout HUG. I just want to make sure the tradeoffs are considered before changing the behavior.

Not sure what you mean by "constraints". The pytest fixture decorator, for example, is optionally callable, and provides a fairly full API of passable kwargs as described in the [API doc here](http://doc.pytest.org/en/latest/reference.html#pytest-fixture). Although we wouldn't necessarily have to implement it the same way as pytest.

Ideally this HOPE would be a specification of the exact API we'd like to implement.

I'm all for exploring implications of our APIs, and I'm not sure what Tim's timeline is, but I think there's plenty of time to explore ramifications of these HOPE docs. Is there a specific experiment you want to run, or a specific question you want to ask?

**Comment by timothycrosley** [comment]
IMHO, This is a pretty awesome discussion! I think in any case, this is going to result in a much better and more thought out implementation of this feature in Hug than would occur otherwise. 

Some things I think should change no matter what based on this pull-request and current state of discussion:

- Instead of `provides` and `requires`, I agree with the more active form: `provide` and `require`.
- Instead of marking `require` using a string, attribute-based access will be used:
          ```
               hug.require("mysql") -> hug.require.mysql
          ```
  This would, I believe resolve both of these:

> `* get rid of string value connascence
also, in contrast to string connascence, we could get IDE tab over completion (implementation and IDE dependent)

* would have functional closure to accommodate monkey patching on dynamic changes. The string default form evaluates on import time only.`

## I do, however, have some reservations about the decorator based approach upon further reflection:

1. What do we do when there is a default specified on the endpoint? 
 With the current proposal, we side-step this consideration as the `require` call replaces the default.
 
2.  It couples the name of the provider to the argument of the function

In the current proposal something like the following would be possible:

```python
@hug.http()
def my_api_call_with_a_lot_cache_calls(cache=require.redis):
    cache.set()
    ....
    cache.get()
    ...

```

3. It removes the ability to easily reuse a requirement multiple times:

```python
@hug.http()
def customer_purchases(customer_id: int,
                                         customers=require.mysql(table="customers"),
                                         purchases=require.mysql(table="purchases")):
    return purchases.where(account_id=customers.get(id=customer_id).account_id)
```

4. If nothing else, it introduces a new requirement always to keep two separately located lines of code in-sync. One of the reasons I love Python's new type annotations is they remove the common case where argument types are specified in the docstring. This change eliminates the repetition that inevitably leads to invalid documentation.

5. The default argument approach follows a pattern used by newer HTTP frameworks who've had to consider this problem recently, and in the same essential capacity. 

6. It's more verbose, especially for more straightforward cases where some customization is wanted but only of one or two parameters, or of aliasing.

## On the other hand there are some things I like about the decorator approach:

1. It keeps the function signature unchanged normal Python. This is useful, as by default hug never changes the given function. The `require` as a default argument would leak into the normal function call, where it likely will not be useful as a default.
2. It would encourage users to create new providers for reuse, which is a better pattern for reuse throughout a project or API:

```python
@hug.provide()
@hug.require.redis()
def cache(redis):
    return redis


@hug.http()
@hug.require.cache()
def my_api_call_with_a_lot_cache_calls(cache):
    cache.set()
```
3. As mentioned, I do agree it would more likely point the user in the right direction the first time they encountered it.
4. I'm sure I'm missing things please feel free to add more.

I am hoping for some additional perspective on these thoughts. In any case, if we choose to go with decorators, we will need to add to the document how we want defaults to behave. We will need to do this both in the case where the function is used locally within Python and in the case where it is accessed from an interface (such as HTTP). I believe both are relatively similar in their implementation complexity.

**Comment by jay-tyler** [comment]
@timothycrosley 

The decorator approach has the advantage of allowing callable code to be run at both import time and on any function call. The default value pattern is only going to execute at import time, but I suppose if you returned a callable as the default value, that could be called on a function call. Although there would be a question of who is doing the calling? Could be the user, some other decorator, or maybe some other point in the code. Not sure how/if the default value pattern works at all without returning a callable.

I think #1 and #4 of your "reservations" could be managed fairly well by checking the signature, via something equivalent to inspect.signature. An exception could be raised at import time if the arg of the correct name isn't there, along with a helpful message for the end user.

#3 is sort of odd. I would have considered data unpacking in a signature to be bad coding style, but I suppose it's a means of taking advantage of type checking. But there's another wrinkle in that to use a provided function, e.g. call something like `cache`, then the default value pattern needs to evaluate to a callable in the signature. I'm not sure there's as much value in type checking a callable, and it looks like [tools like mypy have trouble when a callable is set as a default arg](https://mypy.readthedocs.io/en/latest/kinds_of_types.html#callable-types-and-lambdas).

#5 I'd really like to see to get a sense of it. If you have an API you could point me to, I'll take a look.

#6 of the reservations isn't exactly true. I think in terms of number of characters, the two approaches are nearly the same. But the decorator pattern will use decorator lines to declare dependencies. The default arg pattern will move verbosity to the argument spec.

Also, a totally new point, but I don't know that "require" and "provide" are the best names.
Provide makes sense on the provider end, but if you think about it, either require or provide could make sense on the end being "provided". I think this will lead to some confusion and difficulty remembering the api. 

What about a noun/verb combo, where the provider is wrapped in a noun, and is provided via a verb. Even `provider` and `provide` could be an okay combination, but perhaps there's a better name?

File: all/HOPE-3--Dependency-Injection.md (hunk 4 above)
Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:64]: By constraints I mean you have to leave yourself some way to distinguish a decorator call that is receiving the wrapped function and a decorator that is being called directly by the programmer (that needs to return a function that will wrap the the called function). The links above show some of the ways that you can tell between those use cases. Pytest for example makes sure that the first parameter can only be a string when it is being called by a programmer so that that it can distinguish when the first parameter is instead a `callable`, aka the function to wrap.

```python
def decorator(function_or_string="default", opt_param1=None, opt_param2=42):
   pass
```

```python
@decorator
def function():
```
Here `decorator` is being called with `function` being passed into the first parameter (`function_or_string`). Pytest checks for this by seeing if `callable(function_or_string)` as well as verifying that some of the optional parameters are their default values. In this case you need to return the wrapped function directly.

```python
@decorator()
def function():
```
Here instead `decorator` is being passed no parameters so `function_or_string` instead receives `default`. In this case you need to return a function to return the wrapped function.

The two different ways of executing `decorator` result in different behaviors that you have to have additional logic to detect and behave properly. It is definitely doable, but to make sure you can detect which way you are being executed as a decorator means you have to constrain your parameter types. Specifically you can't accept a function (or class with `def __call__()`) as the first parameter because then you can't distinguish the way in which you are being called.

Comment (jay-tyler) [reply all/HOPE-3--Dependency-Injection.md:64]: Okay, that makes sense. To summarize, we'd have trouble with any kind of decorator that accepted a single positional arg, with that arg being a callable.

e.g. This probably couldn't be implemented

```
@my_optionally_callable_decorator(some_random_callable) 
def wrapped():
    pass
```

Comment (Peter200lx) [reply all/HOPE-3--Dependency-Injection.md:64]: Yup, and there's additional work if you accept multiple parameters and want to make sure they are all in their default state. By no means impossible, but something that we want to go into with eyes open if we're going to make all of hug's decorators optionally callable (and the more decorators support that workflow the more confusion if some decorators don't support it).
